### ISSUE TITLE
mono fails to build on aarch64-darwin

### ISSUE THREAD

**Issue by prusnak** [NixOS/nixpkgs#151720]
Build ends up with

```
if test -w /private/tmp/nix-build-mono-6.12.0.122.drv-1/mono-6.12.0.122/mcs; then :; else chmod -R +w /private/tmp/nix-build-mono-6.12.0.122.drv-1/mono-6.12.0.122/mcs; fi
cd /private/tmp/nix-build-mono-6.12.0.122.drv-1/mono-6.12.0.122/mcs && make --no-print-directory -s NO_DIR_CHECK=1 PROFILES='binary_reference_assemblies net_4_x xbuild_12 xbuild_14              ' CC='clang' all-profiles
mkdir -p -- build/deps
/nix/store/7aqi1cxp0q594zl71c8fpkv7zjnzf6rw-bash-5.1-p12/bin/bash: line 1: mono: command not found
make[6]: *** [build/profiles/build.make:134: build/deps/basic-profile-check.exe] Error 127
*** The runtime 'mono' doesn't appear to be usable.
*** Trying the 'monolite-macos/1A5E0066-58DC-428A-B21C-0AD6CDAE2789' directory.
make[8]: *** [build/profiles/build.make:134: build/deps/basic-profile-check.exe] Segmentation fault: 11
*** The contents of your 'monolite-macos/1A5E0066-58DC-428A-B21C-0AD6CDAE2789' directory may be out-of-date
*** You may want to try 'make get-monolite-latest'
make[8]: *** [build/profiles/build.make:117: do-profile-check-monolite] Error 1
make[7]: *** [build/profiles/build.make:93: do-profile-check] Error 2
make[6]: *** [build/profiles/build.make:129: do-profile-check-monolite] Error 2
make[5]: *** [build/profiles/build.make:93: do-profile-check] Error 2
make[4]: *** [Makefile:58: profile-do--build--all] Error 2
make[3]: *** [Makefile:54: profiles-do--all] Error 2
make[2]: *** [Makefile:706: all-mcs] Error 2
make[2]: Leaving directory '/private/tmp/nix-build-mono-6.12.0.122.drv-1/mono-6.12.0.122/runtime'
make[1]: *** [Makefile:601: all-recursive] Error 1
make[1]: Leaving directory '/private/tmp/nix-build-mono-6.12.0.122.drv-1/mono-6.12.0.122'
make: *** [Makefile:529: all] Error 2
error: builder for '/nix/store/63g0r6i4gililac2ai7fnvac7b4kc1l8-mono-6.12.0.122.drv' failed with exit code 2
```

Full log: https://logs.nix.ci/?key=nixos/nixpkgs.151468&attempt_id=199e4e63-2c40-469e-8fe5-d8e8b73b5043

**Comment by prusnak** [comment]
cc @thoughtpolice @vrthra @obadz

**Comment by prusnak** [comment]
cc @thoughtpolice @vrthra @obadz 

cc @NixOS/darwin-maintainers

**Comment by starcraft66** [comment]
I took a look at this a few weeks ago actually and it left me scratching my head because the log looks pretty much identical to the x86_64-darwin one except that some executables are missing. I'll have to add some debug logging to figure out what's going on.

**Comment by alcarithemad** [comment]
I tried poked at this in nix-shell for a bit and it looks like the bootstrap compiler in `monolite-macos` segfaults on aarch64. I tried installing the current stable Mono binary release* and telling the build to bootstrap from that, and it gets further but eventually segfaults here:

```
$ make V=1 EXTERNAL_RUNTIME=/Library/Frameworks/Mono.framework/Commands/mono EXTERNAL_MCS=/Library/Frameworks/Mono.framework/Commands/mcs
...
make[7]: Leaving directory '/Users/csv/scratch/2022-01-26-monobuild-KLaIPv/mono-6.12.0.122/mono-6.12.0.122/mcs/class/Facades'
make all-local
make[6]: Leaving directory '/Users/csv/scratch/2022-01-26-monobuild-KLaIPv/mono-6.12.0.122/mono-6.12.0.122/mcs/class'
make[6]: Entering directory '/Users/csv/scratch/2022-01-26-monobuild-KLaIPv/mono-6.12.0.122/mono-6.12.0.122/mcs/class/aot-compiler'
make all-local
make[7]: Entering directory '/Users/csv/scratch/2022-01-26-monobuild-KLaIPv/mono-6.12.0.122/mono-6.12.0.122/mcs/class/aot-compiler'
MONO_PATH='./../../class/lib/build-macos/' > build-macos_aot.log 2>&1 /Users/csv/scratch/2022-01-26-monobuild-KLaIPv/mono-6.12.0.122/mono-6.12.0.122/runtime/mono-wrapper --aot=bind-to-runtime-version --debug ./../../class/lib/build-macos//mscorlib.dll || (cat build-macos_aot.log; exit 1)
/bin/sh: line 1: 35347 Segmentation fault: 11  MONO_PATH='./../../class/lib/build-macos/' /Users/csv/scratch/2022-01-26-monobuild-KLaIPv/mono-6.12.0.122/mono-6.12.0.122/runtime/mono-wrapper --aot=bind-to-runtime-version --debug ./../../class/lib/build-macos//mscorlib.dll > build-macos_aot.log 2>&1
make[7]: *** [Makefile:61: ../../class/lib/build-macos//mscorlib.dll.dylib] Error 1
make[7]: Leaving directory '/Users/csv/scratch/2022-01-26-monobuild-KLaIPv/mono-6.12.0.122/mono-6.12.0.122/mcs/class/aot-compiler'
```

The mono-wrapper itself will run and spit out mono's help text, but I'm not sure where to go from here in debugging the issue.

* release 6.12.0.122, which despite coming in a pkg labeled `universal` contains only an x86-64 mono binary, which seems to fail the definition of any iteration of Apple's "universal" releases (some other files in the release are fat binaries with 32 bit code).

**Comment by alcarithemad** [comment]
Turns out upstream fixed this in 6.12.0.129 (nearly a year ago), but hasn't made a stable release since .122.

The current preview channel release (6.12.0.168) builds just fine (and runs at least the simple program I wanted mono for).

Nixpkgs could use the preview channel on darwin (though note that the URL below does not match the pattern in `mono/generic.nix`).

```nix
{ pkgs ? import <nixpkgs> {
    overlays = [
      (self: super: {
        mono6 = super.mono6.overrideAttrs (old: {
          version = "6.12.0.168";
          src = super.fetchurl {
            sha256 = "sha256:1b0hhd6a0l0b62x601ics8m2fnvwknnkj8nn96v3gxymng5r3y65";
            url = "https://download.mono-project.com/sources/mono/preview/mono-6.12.0.168.tar.xz";
          };
        });
      })
    ];
  }
}:
pkgs.mkShell {
  buildInputs = [
    pkgs.mono6
  ];
}
```

**Comment by jiegec** [comment]
The latest Stable Mono release is: 6.12.0 Stable (6.12.0.182). We can bump the version to fix this.

**Comment by jiegec** [comment]
Fixed by pr #181051
